crates/libm-test/src/num.rs

```diff
@@ -0,0 +1,447 @@
+            match bits.checked_add(n) {
+                Some(v) if v > ninf_bits => ninf_bits,
+                Some(v) => v,
+                None => ninf_bits,
+            }
+        };
+        Self::from_bits(next_bits)
+    }
+}
+
+impl<F> FloatExt for F where F: Float {}
+
+/// Extra constants that are useful for tests.
+#[derive(Debug, Clone, Copy)]
+pub struct Consts<F> {
+    /// The default quiet NaN, which is also the minimum quiet NaN.
+    pub pos_nan: F,
+    /// The default quiet NaN with negative sign.
+    pub neg_nan: F,
+    /// NaN with maximum (unsigned) significand to be a quiet NaN. The significand is saturated.
+    pub max_qnan: F,
+    /// NaN with minimum (unsigned) significand to be a signaling NaN.
+    pub min_snan: F,
+    /// NaN with maximum (unsigned) significand to be a signaling NaN.
+    pub max_snan: F,
+    pub neg_max_qnan: F,
+    pub neg_min_snan: F,
+    pub neg_max_snan: F,
+}
+
+impl<F: FloatExt> Consts<F> {
+    fn new() -> Self {
+        let top_sigbit_mask = F::Int::ONE << (F::SIG_BITS - 1);
+        let pos_nan = F::EXP_MASK | top_sigbit_mask;
+        let max_qnan = F::EXP_MASK | F::SIG_MASK;
+        let min_snan = F::EXP_MASK | F::Int::ONE;
+        let max_snan = (F::EXP_MASK | F::SIG_MASK) ^ top_sigbit_mask;
+
+        let neg_nan = pos_nan | F::SIGN_MASK;
+        let neg_max_qnan = max_qnan | F::SIGN_MASK;
+        let neg_min_snan = min_snan | F::SIGN_MASK;
+        let neg_max_snan = max_snan | F::SIGN_MASK;
+
+        Self {
+            pos_nan: F::from_bits(pos_nan),
+            neg_nan: F::from_bits(neg_nan),
+            max_qnan: F::from_bits(max_qnan),
+            min_snan: F::from_bits(min_snan),
+            max_snan: F::from_bits(max_snan),
+            neg_max_qnan: F::from_bits(neg_max_qnan),
+            neg_min_snan: F::from_bits(neg_min_snan),
+            neg_max_snan: F::from_bits(neg_max_snan),
+        }
+    }
+
+    pub fn iter(self) -> impl Iterator<Item = F> {
+        let Self {
+            pos_nan,
+            neg_nan,
+            max_qnan,
+            min_snan,
+            max_snan,
+            neg_max_qnan,
+            neg_min_snan,
+            neg_max_snan,
+        } = self;
+
+        [pos_nan, neg_nan, max_qnan, min_snan, max_snan, neg_max_qnan, neg_min_snan, neg_max_snan]
+            .into_iter()
+    }
+}
+
+/// Return the number of steps between two floats, returning `None` if either input is NaN.
+///
+/// This is the number of steps needed for `n_up` or `n_down` to step between values.
+/// Infinities are treated the same. It does not matter which value is greater.
+pub fn ulp_between<F: Float>(x: F, y: F) -> Option<F::Int> {
+    if x.is_nan() || y.is_nan() {
+        return None;
+    }
+
+    let x_abs = x.abs().to_bits();
+    let y_abs = y.abs().to_bits();
+
+    if x.signum() == y.signum() {
+        Some(max(x_abs, y_abs) - min(x_abs, y_abs))
+    } else {
+        Some(x_abs + y_abs)
+    }
+}
```

Comment:
Could extract a helper, which might also be useful for some of the other functions in this module:
```suggestion
pub fn ulp_between<F: Float>(x: F, y: F) -> Option<F::Int> {
    let a = as_ulp_steps(x)?;
    let b = as_ulp_steps(y)?;

    Some(a.abs_diff(b))
}
/// Return the (signed) number of steps from zero to x
fn as_ulp_steps(x: f32) -> Option<F::SignedInt> {
    let s = x.to_bits_signed();

    let val = if s >= 0 {
        // each increment from `s = 0`
        // is one step up from `x = 0.0`
        s
    } else {
        // each increment from `s = F::SignedInt::MIN`
        // is one step down from `x = -0.0`
        F::SignedInt::MIN - s
    };

    if (s as F::Int & !F::SIGN_MASK) <= F::EXP_MASK {
        Some(val)
    } else {
        // x is NaN
        None
    }
}
```

Reply:
This wasn't causing the tests to fail and I couldn't figure out why, turns out you found a bug in `abs_diff` :) https://github.com/rust-lang/libm/pull/366. Updated with a small adjustment from your suggestion.